saq/queue/postgres.py

```diff
@@ -0,0 +1,608 @@
+                    SQL(
+                        """
+                    DELETE FROM {jobs_table}
+                    WHERE key = %(key)s
+                    """
+                    ).format(jobs_table=self.jobs_table),
+                    {"key": key},
+                )
+        await self._release_job(key)
+
+        self._update_stats(status)
+
+        await self.notify(job)
+        logger.info("Finished %s", job.info(logger.isEnabledFor(logging.DEBUG)))
+
+    async def dequeue(self, timeout: float = 0) -> Job | None:
+        """Wait on `self.cond` to dequeue.
+
+        Retries indefinitely until a job is available or times out.
+        """
+        self.waiting += 1
+        async with self.cond:
+            self.cond.notify(1)
+
+        try:
+            return await asyncio.wait_for(self.queue.get(), timeout or None)
+        except asyncio.exceptions.TimeoutError:
+            return None
+        finally:
+            self.waiting -= 1
+
+    async def wait_for_job(self) -> None:
+        while True:
+            async with self.cond:
+                await self.cond.wait()
+
+            for job in await self._dequeue():
+                await self.queue.put(job)
+
+    async def enqueue(self, job_or_func: str | Job, **kwargs: t.Any) -> Job | None:
+        job = self._create_job_for_enqueue(job_or_func, **kwargs)
+
+        await self._before_enqueue(job)
+
+        async with self.pool.connection() as conn, conn.cursor() as cursor:
+            await cursor.execute(
+                SQL(
+                    """
+                INSERT INTO {jobs_table} (key, job, queue, status, scheduled)
+                VALUES (%(key)s, %(job)s, %(queue)s, %(status)s, %(scheduled)s)
+                ON CONFLICT (key) DO UPDATE
+                SET job = %(job)s, queue = %(queue)s, status = %(status)s, scheduled = %(scheduled)s
+                WHERE {jobs_table}.status NOT IN ('queued', 'aborting')
+                RETURNING job
+                """
+                ).format(jobs_table=self.jobs_table),
+                {
+                    "key": job.key,
+                    "job": self.serialize(job),
+                    "queue": self.name,
+                    "status": job.status,
+                    "scheduled": job.scheduled or seconds(now()),
+                },
+            )
+
+            if not await cursor.fetchone():
+                return None
+
+            await cursor.execute(f"NOTIFY \"{ENQUEUE_CHANNEL}\", '{job.key}'")
+
+        logger.info("Enqueuing %s", job.info(logger.isEnabledFor(logging.DEBUG)))
+        return job
+
+    async def get_abort_errors(self, jobs: Iterable[Job]) -> list[bytes | None]:
+        async with self.pool.connection() as conn, conn.cursor() as cursor:
+            await cursor.execute(
+                SQL(
+                    """
+                SELECT key, job->'error'
+                FROM {jobs_table}
+                WHERE key = ANY(%(keys)s)
```

Comment:
is this always sorted automatically? does the order matter?

Reply:
Yes, order does matter. No, I don't think order is guaranteed here. There's additional logic below to reorder the errors to match the input jobs list.